File: MbitMore.cpp
#include "pxt.h"

#include "MicroBit.h"
#include "MicroBitConfig.h"

#include "MbitMoreCommon.h"

#define UPDATE_PERIOD 19
#define NOTIFY_PERIOD 101

#if MICROBIT_CODAL
#include "MbitMoreService.h"
#else // NOT MICROBIT_CODAL
#include "MbitMoreServiceDAL.h"
using MbitMoreService = MbitMoreServiceDAL;
#endif // NOT MICROBIT_CODAL

//% color=#FF9900 weight=95 icon="\uf1b0"
namespace MbitMore {
  MbitMoreService *_pService = NULL;
#if MICROBIT_CODAL
#else // NOT MICROBIT_CODAL
  int dummyDataLabelID = 0;
#endif // NOT MICROBIT_CODAL

  void update() {
    while (NULL != _pService) {
      _pService->update();
      fiber_sleep(UPDATE_PERIOD);
    }
  }

  void notifyScratch() {
    while (NULL != _pService) {
      // notyfy data to Scratch
      _pService->notify();
      fiber_sleep(NOTIFY_PERIOD);
    }
  }

  /**
   * @brief Start Microbit More service.
   * 
   */
  //%
  void startMbitMoreService() {
    if (NULL != _pService)
      return;

    _pService = new MbitMoreService();

    create_fiber(update);
    // create_fiber(notifyScratch);
  }

  /**
   * @brief Register a label in waiting data list and return an ID for the label.
   * This starts Microbit More service if it was not available.
   * 
   * @param dataLabel label to register
   * @param dataType type of the data to be received
   * @return int ID for the label
   */
  //%
  int call_registerWaitingDataLabel(String dataLabel, MbitMoreDataContentType dataType) {
#if MICROBIT_CODAL
    if (NULL == _pService)
      startMbitMoreService();

    int labelID = _pService->registerWaitingDataLabel(MSTR(dataLabel), dataType);
    return labelID;
#else // NOT MICROBIT_CODAL
    return ++dummyDataLabelID; // dummy
#endif // NOT MICROBIT_CODAL
  }

  /**
   * @brief Get number which was received with the label.
   * 
   * @param labelID ID in registered labels
   * @return float received data with the label
   */
  //%
  float call_dataContentAsNumber(int labelID) {
#if MICROBIT_CODAL
    return _pService->dataContentAsNumber(labelID);
#else // NOT MICROBIT_CODAL
    return 0.0; // dummy
#endif // NOT MICROBIT_CODAL
  }

  /**
   * @brief Get text which was received with the label.
   * 
   * @param labelID ID in registered labels
   * @return String received data with the label
   */
  //%
  String call_dataContentAsText(int labelID) {
#if MICROBIT_CODAL
    return PSTR(_pService->dataContentAsText(labelID));
#else // NOT MICROBIT_CODAL
    return String(""); // dummy
#endif // NOT MICROBIT_CODAL
  }

  /**
   * @brief Send a float with labele to Scratch.
   * Do nothing if Scratch was not connected.
   * 
   * @param dataLabel - label of the data
   * @param dataContent - content of the data
   */
  //%
  void call_sendNumberWithLabel(String dataLabel, float dataContent) {
#if MICROBIT_CODAL
    if (NULL == _pService)
      return;

    _pService->sendNumberWithLabel(MSTR(dataLabel), dataContent);
#endif // MICROBIT_CODAL
  }

  /**
   * @brief Send a text with label to Scratch.
   * Do nothing if Scratch was not connected.
   * 
   * @param dataLabel - label of the data
   * @param dataContent - content of the data
   */
  //%
  void call_sendTextWithLabel(String dataLabel, String dataContent) {
#if MICROBIT_CODAL
    if (NULL == _pService)
      return;

    _pService->sendTextWithLabel(MSTR(dataLabel), MSTR(dataContent));
#endif // MICROBIT_CODAL
  }

} // namespace MbitMore

File: MbitMore.ts
namespace MbitMore {
  const MBIT_MORE_DATA_RECEIVED = 8000;

  /**
  * Starts BLE services for Scratch Microbit-More extension.
  */
  //% blockId=MbitMore_startMbitMoreService block="start Microbit-More service"
  //% shim=MbitMore::startMbitMoreService
  export function startService(): void {
    console.log("Microbit-More started");
  }

  /**
  * Register a label and return its ID.
  */
  //% shim=MbitMore::call_registerWaitingDataLabel
  export function registerWaitingDataLabel(label: string, type: MbitMoreDataContentType): number {
    console.log("Microbit-More registered label: " + label);
    return 1; // dummy for sim
  }

  /**
   * Read received data as a number
   */
  //% shim=MbitMore::call_dataContentAsNumber
  export function dataContentAsNumber(labelID: number): number {
    return 0.0; // dummy for sim
  }

  /**
   * Read received data as a text
   */
  //% shim=MbitMore::call_dataContentAsText
  export function dataContentAsText(labelID: number): string {
    return "text"; // dummy for sim
  }

  /**
   * Run blocks with data when a number data with the label is received.
   * @param label - label of the data
   * @param handler - blocks to run
   */
  //% blockId=MbitMore_onReceivedNumberWithLabel
  //% block="on number $numberData with label $label"
  //% label.defl="label-01"
  //% draggableParameters
  export function onReceivedNumberWithLabel(label: string, handler: (numberData: number) => void) {
    let labelID = MbitMore.registerWaitingDataLabel(label, MbitMoreDataContentType.MM_DATA_NUMBER);
    if (0 === labelID) {
      throw "max waiting label counts exceed";
    }
    control.onEvent(MBIT_MORE_DATA_RECEIVED, labelID, function () {
      handler(MbitMore.dataContentAsNumber(labelID));
      return;
    });
  }

  /**
   * Run blocks with data when a text data with the label is received.
   * @param label - label of the data
   * @param handler - blocks to run
   */
  //% blockId=MbitMore_onReceivedTextWithLabel
  //% block="on text $textData with label $label"
  //% label.defl="label-01"
  //% draggableParameters
  export function onReceivedTextWithLabel(label: string, handler: (textData: string) => void) {
    let labelID = MbitMore.registerWaitingDataLabel(label, MbitMoreDataContentType.MM_DATA_TEXT);
    if (0 === labelID) {
      throw "max waiting label counts exceed";
    }
    control.onEvent(MBIT_MORE_DATA_RECEIVED, labelID, function () {
      handler(MbitMore.dataContentAsText(labelID));
      return;
    });
  }

  /**
   * Send number with label
   * @param label lavel of the data 
   * @param data number value to send
   */
  //% blockId=MbitMore_sendNumberWithLabel
  //% block="send number $numberData with label $label"
  //% shim=MbitMore::call_sendNumberWithLabel
  //% label.defl="label-01"
  //% numberData.defl=0.0
  export function sendNumberWithLabel(label: string, numberData: number): void {
    console.log("Microbit-More send a number: " + label + " = " + numberData);
  }

  /**
   * Send text with label
   * @param label lavel of the data
   * @param data text to send
   */
  //% blockId=MbitMore_sendTextWithLabel
  //% block="send text $textData with label $label"
  //% shim=MbitMore::call_sendTextWithLabel
  //% label.defl="label-01"
  //% textData.defl="text"
  export function sendTextWithLabel(label: string, textData: string): void {
    console.log("Microbit-More send a text: " + label + " = " + textData);
  }

} // namespace MbitMore